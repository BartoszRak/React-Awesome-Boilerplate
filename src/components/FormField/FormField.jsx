import React from 'react'
import PropTypes from 'prop-types'
import { FastField, Field } from 'formik'

import TextField from '@material-ui/core/TextField'

export class FormField extends React.Component {
  static propTypes = {
    checkable: PropTypes.bool,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    component: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
    customOnChange: PropTypes.func,
    fast: PropTypes.bool,
    name: PropTypes.string.isRequired,
  }

  static defaultProps = {
    children: () => null,
    component: TextField,
    customOnChange: () => null,
  }

  renderInput = ({ field, form }) => {
    const {
      checkable,
      children,
      component: Component,
      customOnChange,
      fast,
      name,
      ...inputProps
    } = this.props
    const errors = form.touched[name] && form.errors[name]
    const helperText = inputProps.helperText || ''

    if (checkable) {
      inputProps.checked = field.value === true || field.value === 'true'
      inputProps.value = 'checked'
    } else {
      inputProps.error = !!errors
      inputProps.helperText = errors || helperText
    }

    return (
      <Component
        {...field}
        {...inputProps}
        onChange={evtOrValue => {
          if (typeof evtOrValue === 'string') {
            customOnChange(name, evtOrValue)
            field.onChange(name)(evtOrValue)
          } else {
            customOnChange(evtOrValue)
            field.onChange(evtOrValue)
          }
        }}
      >
        {children(field)}
      </Component>
    )
  }

  render() {
    const { fast, name } = this.props

    const FieldComponent = fast ? FastField : Field

    return <FieldComponent name={name} render={this.renderInput} />
  }
}

export default FormField
