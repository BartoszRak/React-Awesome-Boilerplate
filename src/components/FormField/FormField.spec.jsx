import React from 'react'
import { shallow } from 'enzyme'

import { FormField } from './FormField'

const fakeComponent = ({ children }) => children
const initComponent = overrides => {
  const mockProps = {
    fast: false,
  }
  const wrapper = shallow(<FormField
    component={fakeComponent}
    name="my-field"
    {...mockProps}
    {...overrides}
  />)
  return { mockProps, wrapper }
}

describe('<FormField/>', () => {
  describe('rendering', () => {
    it('renders without crashing', () => {
      const { wrapper } = initComponent()
      expect(wrapper).toBeTruthy()
    })
    it('should render as expected', () => {
      const { wrapper } = initComponent()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('renderInput', () => {
    const { wrapper } = initComponent()
    const instance = wrapper.instance()
    const Input = instance.renderInput

    it('should render provided component and pass input props', () => {
      wrapper.setProps({ helperText: 'Helper Text' })
      const field = { value: 'some value' }
      const form = { errors: {}, touched: {} }
      const renderedInput = shallow(<Input field={field} form={form} />)

      expect(renderedInput.props()).toMatchObject({
        error: false,
        helperText: 'Helper Text',
        value: 'some value',
      })
    })

    it('should render error in helper text only if the field was touched', () => {
      const field = { value: 'some value' }
      let form = { errors: {}, touched: {} }

      let renderedInput = shallow(<Input field={field} form={form} />)
      expect(renderedInput.props()).toMatchObject({
        error: false,
        helperText: 'Helper Text',
      })

      form = {
        errors: { 'my-field': 'Some error' },
        touched: { 'my-field': true },
      }
      renderedInput = shallow(<Input field={field} form={form} />)
      expect(renderedInput.props()).toMatchObject({
        error: true,
        helperText: 'Some error',
      })
    })

    it('should render helper text if it was provided and there are no errors', () => {
      wrapper.setProps({ helperText: 'Helper Text' })
      const field = { value: 'some value' }
      const form = { errors: {}, touched: {} }

      let renderedInput = shallow(<Input field={field} form={form} />)
      expect(renderedInput.props()).toMatchObject({
        error: false,
        helperText: 'Helper Text',
      })

      wrapper.setProps({ helperText: undefined })
      renderedInput = shallow(<Input field={field} form={form} />)
      expect(renderedInput.props()).toMatchObject({
        error: false,
        helperText: '',
      })
    })
  })
})
