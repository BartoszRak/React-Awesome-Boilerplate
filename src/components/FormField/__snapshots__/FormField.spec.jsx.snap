// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField/> rendering should render as expected 1`] = `
<FormikConnect(FieldInner)
  name="my-field"
  render={[Function]}
/>
`;
